platforms/core-configuration/configuration-cache/src/main/kotlin/org/gradle/internal/cc/impl/promo/ConfigurationCachePromoHandler.kt

```diff
@@ -32,24 +35,26 @@ import org.gradle.internal.configuration.problems.PropertyProblem
 import org.gradle.internal.configuration.problems.PropertyTrace
 import org.gradle.internal.configuration.problems.StructuredMessageBuilder
 import org.gradle.internal.deprecation.DeprecationLogger
+import org.gradle.internal.extensions.core.serviceOf
 import org.gradle.internal.service.scopes.Scope
 import org.gradle.internal.service.scopes.ServiceScope
 
 /**
  * This class handles configuration cache promo message at the end of the build.
- * <p>
+ *
  * The promo is shown unless:
- * <ul>
- *     <li>An incompatible API is used at configuration time (in other words, if a CC problem would be emitted by it).</li>
- *     <li>A {@code noCompatibleWithConfigurationCache} task is present in the main task graph.</li>
- *     <li>Graceful degradation is requested.</li>
- *     <li>Build fails.</li>
- * </ul>
+ *   - An incompatible API is used at configuration time (in other words, if a CC problem would be emitted by it).
+ *   - A `noCompatibleWithConfigurationCache` task is present in the main task graph.
+ *   - Graceful degradation is requested.
+ *   - Build fails.
+ *   - Build runs without a project (`gradle init` or `gradle help` outside project directory)
  *
  * Instances of this class are thread-safe.
  */
 @ServiceScope(Scope.BuildTree::class)
 internal class ConfigurationCachePromoHandler(
+    private val startParameter: StartParameter,
+    private val builtInCommands: List<BuiltInCommand>,
```

Comment:
💭  I am not a fan of the built-in command list being injected in many places. It feels that many or all injection points use it at the wrong level of abstraction.

If you have spare cycles, it would be nice to introduce a dedicated service that does this kind of checking. Probably best done in a separate PR.

Reply:
> 💭 I am not a fan of the built-in command list being injected in many places. It feels that many or all injection points use it at the wrong level of abstraction.

After thinking more about this, I realized that not having a build definition is already enough signal to suppress - we should only show promo if there is a build to enable CC in.